### ISSUE TITLE
Encrypted files are not self-contained

### ISSUE THREAD

**Issue by C0rby** [owncloud/encryption#211]
Not all information that are relevant for verifying the integrity of the encrypted file blocks are stored within the files. The file versions are stored in the database instead. Due to this you have to backup the database in addition to the encrypted files and the key material to be able to properly decrypt the files again. This design decision can also lead to a loss of integrity when a database restore is required as the file versions within the database may not match the files on disk anymore.

**Comment by mmattel** [comment]
Do you have an idea for a solution ?

**Comment by C0rby** [comment]
One option is to add the file version to the encrypted file header.

**Comment by mmattel** [comment]
Adding @micbar, maybe this of interest

**Comment by mmattel** [comment]
@karakayasemi you have/gained a lot of experiences, maybe you have an idea 😃

**Comment by karakayasemi** [comment]
@mmattel as @C0rby said, we can add this information to header. But, again we need to think about a solution for the old files. 

We can first try to read file version from the header first, if we do not find in there, we can check from database.

**Comment by mmattel** [comment]
good idea, could be combined with #224. When rewritten, add to header, remove from db...

**Comment by C0rby** [comment]
> @mmattel as @C0rby said, we can add this information to header. But, again we need to think about a solution for the old files.
> 
> We can first try to read file version from the header first, if we do not find in there, we can check from database.

I think this is the way to go.

> When rewritten, add to header, remove from db...

The backend still needs the version field in the DB IIRC.

**Comment by karakayasemi** [comment]
> 
> 
> good idea, could be combined with #224. When rewritten, add to header, remove from db...

We can not combine with #224. It needs complete rewrite of the file with binary encoding.

**Comment by mmattel** [comment]
@karakayasemi any news on that?

**Comment by mmattel** [comment]
I regulary ping that issue... 😅
